File: c++/1261.cpp
// 0-1 bfs
#include <deque>
#include <iostream>
#include <string>

using namespace std;

int m, n;
int miro[100][100];
int dy[4] = {1, 0, -1, 0};
int dx[4] = {0, 1, 0, -1};
bool isVisited[100][100];
int crushed[100][100];

bool inRange(int y, int x) { return 0 <= y && y < n && 0 <= x && x < m; }

void bfs() {
  deque<pair<int, int>> dq;

  dq.push_back({0, 0});
  isVisited[0][0] = true;

  while (!dq.empty()) {
    int y = dq.front().first;
    int x = dq.front().second;
    dq.pop_front();

    if (y == n - 1 && x == m - 1) {
      cout << crushed[y][x];
      return;
    }

    for (int i = 0; i < 4; i++) {
      int ny = y + dy[i];
      int nx = x + dx[i];

      if (inRange(ny, nx) && !isVisited[ny][nx]) {
        isVisited[ny][nx] = true;
        if (miro[ny][nx] == 0)
          dq.push_front({ny, nx});
        else
          dq.push_back({ny, nx});
        crushed[ny][nx] = crushed[y][x] + miro[ny][nx];
      }
    }
  }
}

int main() {
  freopen("example.txt", "r", stdin);
  cin >> m >> n;

  for (int i = 0; i < n; i++) {
    string s;
    cin >> s;
    for (int j = 0; j < s.length(); j++) {
      miro[i][j] = s[j] - '0';
    }
  }

  bfs();

  return 0;
}

// dijkstra>, pq
// #include <iostream>
// #include <queue>
// #include <string>
// #include <vector>

// using namespace std;

// int m, n;
// int miro[100][100];
// int dy[4] = {1, 0, -1, 0};
// int dx[4] = {0, 1, 0, -1};
// bool isVisited[100][100];
// int crushed[100][100];

// bool inRange(int y, int x) { return 0 <= y && y < n && 0 <= x && x < m; }

// void dijkstra() {
//   priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int,
//   int>>>,
//                  greater<pair<int, pair<int, int>>>>
//       pq;

//   pq.push({0, {0, 0}});
//   isVisited[0][0] = true;

//   while (!pq.empty()) {
//     int cnt = pq.top().first;
//     int y = pq.top().second.first;
//     int x = pq.top().second.second;
//     pq.pop();

//     if (y == n - 1 && x == m - 1) {
//       cout << cnt;
//       return;
//     }

//     for (int i = 0; i < 4; i++) {
//       int ny = y + dy[i];
//       int nx = x + dx[i];

//       if (inRange(ny, nx) && !isVisited[ny][nx]) {
//         isVisited[ny][nx] = true;
//         if (miro[ny][nx] == 0)
//           pq.push({cnt, {ny, nx}});
//         else
//           pq.push({cnt + 1, {ny, nx}});
//       }
//     }
//   }
// }

// int main() {
//   freopen("example.txt", "r", stdin);
//   cin >> m >> n;

//   for (int i = 0; i < n; i++) {
//     string s;
//     cin >> s;
//     for (int j = 0; j < s.length(); j++) {
//       miro[i][j] = s[j] - '0';
//     }
//   }

//   dijkstra();

//   return 0;
// }

File: js/1261.js
// 길이가 3인 배열을 q에 담기
const fs = require('fs');

const inputs = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const dy = [1, 0, -1, 0];
const dx = [0, 1, 0, -1];

const [m, n] = inputs[0].split(' ').map(Number);
const miro = [];
for (let i = 1; i <= n; i++) {
  const subArr = inputs[i].split('').map(Number);
  miro.push(subArr);
}

const isVisited = Array(n)
  .fill()
  .map(() => Array(m).fill(false));

const dq = [[0, 0, 0]];
isVisited[0][0] = true;
while (dq.length !== 0) {
  const [y, x, cnt] = dq.shift();

  if (y === n - 1 && x === m - 1) {
    console.log(cnt);
    break;
  }

  for (let i = 0; i < 4; i++) {
    const ny = y + dy[i];
    const nx = x + dx[i];
    if (inRange(ny, nx) && !isVisited[ny][nx]) {
      isVisited[ny][nx] = true;
      if (miro[ny][nx] === 0) dq.unshift([ny, nx, cnt]);
      else dq.push([ny, nx, cnt + 1]);
    }
  }
}

function inRange(y, x) {
  return 0 <= y && y < n && 0 <= x && x < m;
}

// crushed 배열 사용
// const fs = require('fs');

// const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

// const dy = [1, 0, -1, 0];
// const dx = [0, 1, 0, -1];

// const [m, n] = inputs[0].split(' ').map(Number);
// const miro = [];
// for (let i = 1; i <= n; i++) {
//   const subArr = inputs[i].split('').map(Number);
//   miro.push(subArr);
// }

// const isVisited = Array(n)
//   .fill()
//   .map(() => Array(m).fill(false));

// const crushed = Array(n)
//   .fill()
//   .map(() => Array(m).fill(0));

// const dq = [[0, 0]];
// isVisited[0][0] = true;
// while (dq.length !== 0) {
//   const [y, x] = dq.shift();

//   if (y === n - 1 && x === m - 1) {
//     console.log(crushed[y][x]);
//     break;
//   }

//   for (let i = 0; i < 4; i++) {
//     const ny = y + dy[i];
//     const nx = x + dx[i];
//     if (inRange(ny, nx) && !isVisited[ny][nx]) {
//       isVisited[ny][nx] = true;
//       if (miro[ny][nx] === 0) dq.unshift([ny, nx]);
//       else dq.push([ny, nx]);
//       crushed[ny][nx] = crushed[y][x] + miro[ny][nx];
//     }
//   }
// }

// function inRange(y, x) {
//   return 0 <= y && y < n && 0 <= x && x < m;
// }
